Init: organization Profile

idlip opened the pull request:
Related to #756 

Very much WIP!

So far has Basic doctype and web page skeleton.

TODO:
- [x] Interconnect user profiles & Events
- [x] Follow fossunited theme and make it look uniform
- [x] Make it unique enough, orgs are very broad...


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Introduced a comprehensive Organization module with detailed fields, web views, and support for social links, media, and member management.
  * Added templates for displaying organization profiles and lists, including sponsored events, hackathons, jobs, and member cards.
  * Enabled linking user profiles to organizations with specific roles.
  * Added new role and identifier constants for organizations and event sponsors.
  * Added a job board component to display active and expired jobs associated with organizations.
  * Added utility to insert test organizations with members and leads for easier testing.

* **Bug Fixes**
  * None.

* **Tests**
  * Added extensive tests covering organization creation, social links, member retrieval, job listings, permissions, and error handling.

* **Chores**
  * Updated modules list to include Organization.

* **Refactor**
  * Replaced the previous minimal Organization schema with an enhanced version, removing deprecated files and logic.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

idlip committed:
init: basic organization doctype skeleton

idlip committed:
add supporting organization team member doctype

+ adds simple doctype for org team members list
+ used as table for team members linking foss user profiles
+ to avoid linting errors commented in org_role item

idlip committed:
auto-gen files for doctype

idlip committed:
initial web page for organization

+ simply lists all organization
+ each page shows details
 - needs improvements in following theme of fossunited

rahulporuri commented:
@idlip please add demo videos/screenshots like you have done in earlier PRs, and tag me for review when you want me to take a look at the PR

idlip commented:
So far the page looks goofy.
Here is initial screenshot.

- Listing page for all orgs
<img width="2555" height="831" alt="image" src="https://github.com/user-attachments/assets/77bc6909-d8db-4927-ad5e-4ee0e08b8ab7" />

- Each org page
<img width="2379" height="1287" alt="image" src="https://github.com/user-attachments/assets/cb0dca1a-ecd5-476d-9a25-33a9a6df2d20" />

rahulporuri commented:
@idlip please don't force push changes to this branch, it'll be great if you can push newer changes as new commits to the same branch.

I'll use "squash-merge" when I merge this PR so even if you push more than one commit to this branch, they will get squashed into a single commit in the develop branch.

rahulporuri commented:
now that we have a skeleton, let's list the people associated with the org + past instances where the org was a "venue sponsor" on the page

rahulporuri commented on fossunited/fossunited/doctype/organization/organization.json:
```diff
@@ -1,65 +1,226 @@
+ "name": "Organization",
+ "naming_rule": "By fieldname",
+ "owner": "Administrator",
+ "permissions": [
+  {
+   "create": 1,
+   "delete": 1,
+   "email": 1,
+   "export": 1,
+   "print": 1,
+   "read": 1,
+   "report": 1,
+   "role": "System Manager",
+   "share": 1,
+   "write": 1
+  },
+  {
+   "create": 1,
+   "email": 1,
+   "export": 1,
+   "print": 1,
+   "read": 1,
+   "report": 1,
+   "role": "All",
+   "select": 1,
+   "share": 1,
+   "write": 1
+  }
```
This is important to understand. Can you look into the permissions for frappe doctypes and update the perms on this doctype accordingly?

idlip replied:
> This is important to understand. Can you look into the permissions for frappe doctypes and update the perms on this doctype accordingly?

Thanks for the nudge. This was for local testing. Will correct it!!!

idlip committed:
make organization into new module

+ Its a big thing and broad

idlip committed:
update permission to org and add a org member role

idlip committed:
update organization team member doctype

- Better to inherit data from FOSS User profile
So only require info is User profile and their Role in org

idlip commented:
> @idlip please don't force push changes to this branch, it'll be great if you can push newer changes as new commits to the same branch.
> 
> I'll use "squash-merge" when I merge this PR so even if you push more than one commit to this branch, they will get squashed into a single commit in the develop branch.

Yes yes for sure. 😱
Prev PRs were small so I did quick force.

For this I very much try to make atomic commits as per small changes.

idlip commented on fossunited/fossunited/doctype/organization/templates/organization.html:
```diff
@@ -0,0 +1,136 @@
+				{% if telegram %}<li><a href="{{ telegram }}" target="_blank">📢 Telegram</a></li>{% endif %}
+				{% if linkedin %}<li><a href="{{ linkedin }}" target="_blank">🔗 LinkedIn</a></li>{% endif %}
+				{% if instagram %}<li><a href="{{ instagram }}" target="_blank">📸 Instagram</a></li>{% endif %}
+				{% if discord %}<li><a href="{{ discord }}" target="_blank">👾 Discord</a></li>{% endif %}
```
I will better impart what Foss chapter does with svg and so.
It was kept as some template, have to do it in python...

idlip committed:
fix coderabbit diagnosis

- fix permission
- fix showing members role and not whole data
- fix svg alt information

idlip committed:
update social handling python code

idlip committed:
Add role column for organization members table

- Annoyed me so much that it needs list view to be checked

idlip committed:
add github social for org

idlip committed:
update social names

idlip committed:
add social links + get members lists for org pages

idlip committed:
refactor org page following fossunited theme

- Follows foss chapter theme itself
- Appears almost similar

idlip committed:
fix member name

- since org team member doctype has only foss profile + role
we need to use profile for names

idlip committed:
add youtube social button

idlip committed:
render markdown via editing it directly

- Add github icon which was missing before
Since it has only _light and _dark variant

idlip committed:
minor fixup on default banner/logo

idlip commented:
More changes yay ;)

Changelog so far:

- Moved organization into its own module - f958b31
- Update permissions and fix coderabbit linting errors/suggestions - 3b62cac, 6045b84
- Update org team member to completely inherit from FOSS user profile - b196e91
- update social svg logo via svgl, follows same pattern as chapters - e78627b
- Add role column for adding org members - 544b1c1
  It definitely annoyed me. Needed to toggle list_view on it.
- Fix github + discourse link  and svg - 6d63337, 544b1c1
- Refactor org page to follow fossunited chapters theme so it looks uniform - 3b58a38
- Add youtube link
- Make org page render raw markdown format into html
  This makes more sense to preview and type, rather than use text-editor components to define each element. - e1ab6bd
- Rebased on develop branch to include and add doctype_ids for Org as well

idlip commented:
Screenshot for sample org page so far:
<img width="1139" height="1529" alt="image" src="https://github.com/user-attachments/assets/71e85dbf-42cb-4eed-810c-91a00ca39a3f" />

idlip commented:
> now that we have a skeleton, let's list the people associated with the org + past instances where the org was a "venue sponsor" on the page

I have few doubts with this.

AFAIU sponsors are mostly added in hackathons and big Meetups.
And its simply produced as a child table there.

I can think of 2 ways to approach this:

1. Follow whatever is already there, and hack the way to get exact sponsor name to match for Organization.
 Eg: https://archive.fossunited.org/fosshack/2020
 For this Red hat is a sponsor, so the full name should match for me to get them listed a past "Sponsors" to events or venue to list them in Org page.

2. Kinda ideal approach
  Make all sponsors associated as an Organization itself. Just like how every member is linked as `FOSS User Profile`, it would make sense to map every sponsors as `Organization` now.
  This IMO would need refactoring to other pages to reflect the updates.

rahulporuri commented:
> Update org team member to completely inherit from FOSS user profile

Do we need to create a new doctype for this? Can't we use the `FOSS User Profile` directly somehow, e.g., maybe an "organization" attribute or something?

idlip commented:
> > Update org team member to completely inherit from FOSS user profile
> 
> Do we need to create a new doctype for this? Can't we use the `FOSS User Profile` directly somehow, e.g., maybe an "organization" attribute or something?

I think they need to be child table to be listed as team members
Also members might need some permission to read+write their org pages.

This is similar to chapters as well. We have `FOSS Chapter Lead Team Member, FOSS Hackathon Team Member, FOSS Global CFP Review Member`

rahulporuri commented:
> I can think of 2 ways to approach this:

I'd go with option 1 instead of 2 for now - but with a small tweak. We have the "FOSS Event Sponsor" doctype so querying the orgname should provide us with information.

Long-term solution is definitely option 2 but we can ignore that for now.

idlip commented:
> Do we need to create a new doctype for this? Can't we use the FOSS User Profile directly somehow, e.g., maybe an "organization" attribute or something?

Ok I see what you mean here...
Definitely adding `Organization` attr to User profile itself might solve this.

But I think the control over organization will be towards each Foss user profile?
They can simply add themselves to any Organization.
And there won't be any simple table for members list in frappe UI.
We should define role for them in organization as well.

idlip committed:
fix coderabbit linting

- as it suggested

rahulporuri commented:
>  Definitely adding Organization attr to User profile itself might solve this.

yeah, long-term we have tradeoffs that we need to evaluate and look at a few implementations to see which is better.

theoretically, here's what I imagine

- org pages can be standalone, without any mention of team/org members
- foss user profile can "place a request" (somehow) to "manage" an org page. Approval could be verify that the user has an "@org.com" email address
- once approved, the foss user profile (or org member) can "manage" the org page, e.g., add links, update description, add other members to the org, etc
- last option is that someone proactively creates an org account e.g. via user dashboard. no approval is required here

integration with the rest of the platform
- org members can submit, update, expire org job applications
- org members can submit a foss pledge on behalf of the org

but this involves a lot of other pieces which are tangential.

For the sake of this PR, I believe the best thing to do next would be to test these pieces thoroughly - could you look at some of the other tests and add the necessary unittests/integration tests? @idlip

harshtandiya commented on fossunited/organization/doctype/organization/organization.json:
```diff
@@ -0,0 +1,249 @@
+   "share": 1,
+   "write": 1
+  },
+  {
+   "read": 1,
+   "role": "All"
+  },
```
Don't need this

idlip replied:
Hi! Got it!

idlip committed:
remove org team member doctype + add org attr to foss user profile

- Remove org team member doctype
  Although might be inconvenience long term
  But org focuses more on events, sponsors, jobs and others.
- Adds organization attr to FOSS user profile so we directly depend on it
  And make it inter-connect

idlip committed:
add org link and role for user profile

- Org role only appears if Organization is chosen

idlip committed:
add foss event sponsor constant to doctype id

idlip committed:
add sponsored events and hackathon card for org page

idlip committed:
add cards for events and hackathon sponsored by the org

- Shows the past and present sponsoring status by the organization
for both events and hackathons

idlip commented:
Changelog further:

- Removed org team member doctype and made it to depend on foss user profile
  - Foss user profile has two new attr, one for choosing Organization, then if chosen there will be Org role option to select
- Now the org page shows the events and hackathons sponsored by them

Here is my video:

https://github.com/user-attachments/assets/67dea14e-98e8-4364-aa87-60b7235a2f0b

idlip committed:
fix coderabbit lint on eval:doc.org_link

idlip committed:
sanitize html content converted via markdown

rahulporuri commented:
> Now the org page shows the events and hackathons sponsored by them

brilliant!

idlip committed:
add jobs card macro file

- helps to define jobs card from any pages
usage: python code must get context for active and past jobs
 job_board(active_jobs, past_jobs)
 where these two are doctype filters fetched with all fields.

idlip committed:
show jobs for org pages

- shows active and past jobs
Note: As of writing it depends that Jobs company_name: must match
org_name

idlip committed:
fetch jobs for organization name

idlip committed:
test: add utility to insert test Organization

- Adds a utility function insert_test_organization to test Organization
- Validates presence of user profiles for org_lead and members.

idlip committed:
add doc to get_org_jobs

idlip committed:
add unit tests for organization doctype

- creation of organizations with lead and member profiles using a utility function.
- verification of linked user profiles and social media fields.
- Testing of methods: get_social_links, get_members, and get_org_jobs.
- validation for error handling on invalid org leads.
- Handling of partial member data without raising exceptions.
- Ensuring correct return values when no sponsored documents exist.

- Note: Commented out jobs part, since it needs Job doctype well defined by code
  We have it in json :-|

idlip commented:
- Now org page also shows Jobs matching `company_name' as `org_name'
- Added unit tests for Organization
  - Added the utility function to create org in `utils.py`, then rest of it in `test_organization.py'
  - creation of organizations with lead and member profiles using a utility function.
  - verification of linked user profiles and social media fields.
  - Testing of methods: get_social_links, get_members, and get_org_jobs.
  - validation for error handling on invalid org leads.
  - Handling of partial member data without raising exceptions.
  - Ensuring correct return values when no sponsored documents exist.
  
```text
  ----------------------------------------------------------------------
Ran 6 tests in 0.530s

OK
```
  
  - Note: I've commented out jobs part, since it needs Job doctype well defined by code
    We have it in json :-|
    
 
 Here is SS showing jobs card.
 
<img width="1167" height="1621" alt="image" src="https://github.com/user-attachments/assets/042bfcb2-0f43-47c7-8a41-ff32cb8f82b4" />

idlip committed:
fixup misc

- commented out another job part

rahulporuri commented:
> Added unit tests for Organization 

this is a great place to start, but I would also like to see permissions-related tests, e.g., admin/system user can make changes to an organization, e.g., update social links, but a normal user will not be able to make changes, organization info can be retrieved by a non-admin user, e.g., not logged in/guest user

idlip committed:
remove unused date field arg

idlip committed:
add more unit tests

- only authorized users (admin, system user, org lead) can edit organization data.
- normal users (non-members or unrelated users) cannot edit.
- anyone, even unauthenticated users, can retrieve public org info like name, description, social links.

idlip commented:
> this is a great place to start, but I would also like to see permissions-related tests, e.g., admin/system user can make changes to an organization, e.g., update social links, but a normal user will not be able to make changes, organization info can be retrieved by a non-admin user, e.g., not logged in/guest user

I have added those permission tests.

idlip committed:
print debug only for developing, not in production or CI

idlip committed:
remove unrelated tests + remove test user record after tests

idlip committed:
fix deleting users

idlip commented:
Closing this myself, as this would be better to start off again when planned.
Now that I've more clarity as well ;- )
